version: '3'
services:
  prometheus:
    image: prom/prometheus
    ports:
      - "9090:9090"
    volumes:
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
    networks:
      - monitoring

  grafana:
    image: grafana/grafana
    ports:
      - "3000:3000"
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=123
    networks:
      - monitoring
    volumes:
      - ./grafana/provisioning/dashboard.yaml:/etc/grafana/provisioning/dashboards/dashboard.yaml
      - ./grafana/provisioning/datasource.yaml:/etc/grafana/provisioning/datasources/datasource.yaml
      - ./grafana/dashboards:/var/lib/grafana/dashboards
      - grafana-data:/var/lib/grafana

  scala-vertx-app:
    build:
      context: ./scala-app
      dockerfile: "./Dockerfile"
    ports:
      - "8080:8080"
    healthcheck:
      test: [ "CMD", "curl", "localhost:8080" ]
      interval: 10s
      timeout: 10s
      retries: 3
      start_period: 10s
    environment:
      - APP_PORT=8080
    networks:
      - monitoring

  scala-vertx-app-load-client:
    build:
      context: ./client
      dockerfile: "./Dockerfile"
    environment:
      - TARGET_APP_HOST=scala-vertx-app
      - TARGET_APP_PORT=8080
      - TARGET_LOAD_RPS=100
    depends_on:
      scala-vertx-app:
        condition: service_healthy
        restart: true
    networks:
      - monitoring

networks:
  monitoring:
    external: true

volumes:
  grafana-data:
